kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncThreadPoolRegistration.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package dev.responsive.kafka.api.async.internals;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class AsyncThreadPoolRegistration {
+
+  private final AsyncThreadPool threadPool;
+
+  private final Map<AsyncProcessorId, Runnable> asyncProcessorsToFlush = new HashMap<>();
+
+  public AsyncThreadPoolRegistration(
+      final AsyncThreadPool threadPool
+  ) {
+    this.threadPool = threadPool;
+  }
+
+  public AsyncThreadPool threadPool() {
+    return threadPool;
+  }
+
+  public void registerAsyncProcessor(final AsyncProcessorId id, final Runnable flushProcessor) {
+    asyncProcessorsToFlush.put(id, flushProcessor);
+  }
+
+  public void unregisterAsyncProcessor(final AsyncProcessorId id) {
+    asyncProcessorsToFlush.remove(id);
+    threadPool.removeProcessor(id);
+  }
+
+  public void flush() {
+    // TODO: this is non-ideal for stateless apps since we're flush each processor instance
+    //  sequentially and might end up waiting for async results from one processor while there
+    //  are others ready and waiting on other processors/partitions.
+    //  We should loop over all processors and only wait/block when we really have nothing to do
+    //  (doesn't matter for stateful apps since this method is a no-op due to flushing
+    //  everything during #flushCache which comes earlier in the commit process)
+    asyncProcessorsToFlush.values().forEach(Runnable::run);
```

Comment:
we need to iterate on these in topology order right? Also, assuming we do that, then what happens if there is a stateless async processor before a stateful one? I think in that case the stateful processor may do some flush work when it's flushed here. Which I think is fine - just making sure I understand.

Reply:
well technically no because these are all the same processor just from different partitions, but yeah we should probably make sure to flush in topological order to prepare for when we support more than one async processor per topology

As for how to deal with mixed stateless and stateful async processors, that's a good question which I'll want to think through more carefully but will punt on that here since it's really part of the "support multiple async processors" work and not in scope here. I will file a ticket or jot down this question to make sure we address it once we get to that 

(my gut reaction is that it's mostly fine since we re-flush the stateful processor here anyway, but we might also need to re-flush the Streams cache after/as part of flushing the stateful processor, since that might send some records downstream)

Reply:
Alright I made sure that we'll now flush processors within a task in topological order (luckily they're initialized in order so we can just throw them in a list). As for processors across tasks, they can go in any order (and eventually we could consider parallelizing these if it becomes a bottleneck)